### PR TITLE
Evasion Rework / Bloodline Bonuses

### PR DESCRIPTION
**Pull request by aumu1031**
[Evasion Changes]

Speed/CSpeed converts to evasion (for their respective offense type) based on ratio to stat total.
- 0% investment = 0% evasion
- 35% investment = 35% evasion
- 50% investment = 50% evasion

Soft cap at 50%
Hard cap at 75%
Soft cap ratio 50% (any evasion higher than the soft cap is only 50% as effective)

Final evasion is calculated by comparing the evasion stat of both players. 
If Player 1 has 35% evasion, and Player 2 has 10% evasion, the end result is Player 1 with 25% evasion.

Evasion calculation factors in the difference between user's and opponent's stat total.
- e.g. Fighting an opponent with half of your stat total doubles your total evasion. 
- if it was based on your own stat total, gaining non-speed stats decreases your evasion
- if it was based on your own stat total, a player with less speed can have a higher evasion than you (and other weirdness)

ON RELEASE:
- Adjust all Speed Nerf jutsu to Evasion Nerf
- Optional: Adjust all Speed Boost jutsu to Evasion Boost
- Adjust AI stats if necessary

===========================================

Baseline for comparison:
Build 1: 50% Bloodline Skill, 50% Offense Skill
Build 2: 32% Bloodline Skill, 32% Offense Skill, 36% Speed Stat
e.g. 125k/125k (or 0/250k) vs 80k/80k/90k

Before Update: 35% Evasion
After Update: 36% Evasion

Typical matchups should see no statistical change.

===========================================

[Jutsu Changes]

Evasion Boost/Nerf add or subtract evasion rather than modify speed.
Jutsu will be updated to reflect the (current live) amount of evasion gained against an identical opponent.

How Evasion Boost/Nerf values below were determined:
e.g. Nerfing an opponent's speed by 10% is equivalent to decreasing their evasion by 5.22%
1000 vs 1000 => 1000 vs 900 with 10% nerf || thus 1000 / 900 = 1.11111 || (1.11111 - 1) * 0.47 (evasion ratio) = 0.0522
e.g. Increasing your own speed by 10% is equivalent to increasing your evasion by 4.7%
1000 vs 1000 => 1100 vs 1000 with 10% boost || thus 1100 / 1000 = 1.1 || (1.1 - 1) * 0.47 (evasion ratio) = 0.047

Effective changes to speed nerf jutsu:
33% Speed Nerf = 17% Evasion Nerf
25% Speed Nerf = 13% Evasion Nerf
18% Speed Nerf = 9% Evasion Nerf
10% Speed Nerf = 5% Evasion Nerf
(Note these values are against an opponent with equal speed, so these are the least impactful use cases)

Revised to even out the numbers and give small positive adjustment:
33% Speed Nerf = 20% Evasion Nerf
25% Speed Nerf = 15% Evasion Nerf
18% Speed Nerf = 10% Evasion Nerf
10% Speed Nerf = 5% Evasion Nerf (this is on a low-rank jutsu, should be replaced with jutsu overhaul)

Speed boost jutsu can either be changed to Evasion Boost, or kept as Speed/CSpeed Boost.
If kept as Speed/CSpeed Boost will be stronger for high speed builds specifically, which is fine. 

====================================

[Bloodline Boost Changes]

Speed/CSpeed: no change, is now the baseline for comparison
~~Damage Resist: new boost type, applies to all damage types, value now the same as Heal~~
Damage Resist: new boost type, applies to all damage types, scales as % similar to evasion
~~Heal: scales based on remaining HP, 25% weaker at full HP, 25% stronger at 0 HP, distinguishes itself from Resist~~
Heal: same as now, value rebalanced relative to speed/resist

Other:
- Decreased Heal/Resist ratio from 0.036 to 0.030 accounting for Speed BL effectiveness change
- Changed Bloodline Defense from 35 => 50 so we don't deal with an extra 0.7x multiplier, and can set Resist = Heal
- Made 100 base bloodline skill only apply when skill less than 100

ON RELEASE:
- Doing the math on 30 Off vs 25 Off BLs shows that 5 Off is worth 10 Spd/Cspd/Resist/Heal
- For all 25 Off BLs, increase defensive boost power from 15 => 20
- Terracotta Soldiers: 30 Off, 15 Tai Resist, 15 Nin Resist => 30 Off, 10 Resist
- Phantasmal Stalker: 30 Off, 15 Tai Resist, 15 Gen Resist => 30 Off, 10 Resist
- Ancient's Deception: 25 Off, 15 Heal => 25 Off 20 Heal
- Eyes of God: 30 Off, 15 Tai Resist, 15 Gen Resist => 30 Off, 10 Resist
- Smouldering Sands: 30 Off, 15 Tai Resist, 15 Nin Resist => 30 Off, 10 Resist
- Frigid Tundra: 25 Off, 15 CSpd => 25 Off, 20 CSpd
- Impregnable Defenses: 30 Off, 15 Tai Resist, 15 Nin Resist => 25 Off, 20 resist
- Godslayer: 30 Off, 15 Tai Resist, 15 Gen Resist => 30 Off, 10 Resist
- Weeping Skies: 25 Off, 15 Spd => 25 Off 20 Spd

======================================

Side note regarding average damage per turn, important for comparing BL bonuses:
- The standard build in the game is a 30 Off boost BL, with 80k BL skill, 80 Off skill, 90k speed
- A 125k/125k 0 speed build is roughly the same in terms of damage output
- Explosive Kunai Seal has a total power of 9.05 at lv100 factoring in Residual, neither high nor low and is the baseline for  Ninjutsu damage/turn
- Explosive Kunai Seal is roughly 73420 damage/use for a standard Nin build, which equates to 367100 damage after 5 uses, or a 6 turn fight with overkill

### CODE REVIEW

File: classes/battle/BattleManager.php [hunk 1]
```diff
@@ -1053,22 +1060,26 @@ public function jutsuCollision(
     /**
      * @throws RuntimeException
      */
-    private function getEvasionStatAmount(Fighter $fighter, Jutsu $fighter_jutsu): float|int {
+    private function getEvasionStatAmount(Fighter $fighter, Jutsu $fighter_jutsu, int $target_stat_total): float|int {
```
Comment (levimeahan) [inline classes/battle/BattleManager.php:1063]: This function is really performing a different purpose now, we should rename
```suggestion
    private function getEvasionPercent(Fighter $fighter, Jutsu $fighter_jutsu, int $target_stat_total): float|int {
```

File: classes/Bloodline.php [hunk 2]
```diff
@@ -15,6 +15,8 @@ class Bloodline {
 
     const BOOST_POWER_PRECISION = 5;
 
+    const HEAL_RANGE_PERCENT = 25;
```
Comment (levimeahan) [inline classes/Bloodline.php:18]: Could you add a comment explaining this? (healing varies from -HEAL_RANGE_PERCENT at 100% health, to +HEAL_RANGE_PERCENT at 0% health)

**Review by levimeahan — COMMENTED** [review]
Approach here looks good. Let's wait to merge so other PRs in the pipeline can go through

File: classes/Bloodline.php (hunk 2 above)
Comment (aumu1031) [reply classes/Bloodline.php:18]: 75% effectiveness at 100% HP, 125% effectiveness at 0% HP

**Comment by tn5421** [comment]
> Approach here looks good. Let's wait to merge so other PRs in the pipeline can go through

What PRs in particular are blocking this one?

**Comment by levimeahan** [comment]
> > Approach here looks good. Let's wait to merge so other PRs in the pipeline can go through
> 
> What PRs in particular are blocking this one?

Want to get #468 and #475 merged and deployed on prod without waiting for this change which we're going to pair with a couple other PRs

File: classes/Bloodline.php (hunk 2 above)
Comment (aumu1031) [reply classes/Bloodline.php]: Reverted Heal to original mechanics.

File: classes/battle/BattleManager.php (hunk 1 above)
Comment (aumu1031) [reply classes/battle/BattleManager.php]: Done.

**Review by levimeahan — APPROVED** [review]

**Pull request merged**